### PR TITLE
cKDTree optimization

### PR DESCRIPTION
**Pull request by redhog**
Closes https://github.com/mmaelicke/scikit-gstat/issues/58

### CODE REVIEW

File: skgstat/Kriging.py
Comment (redhog) [inline skgstat/Kriging.py:399]: For non-sparse datasets this might actually be a performance bottleneck, and we should use toarray() or somesuch solution to speed it up.

**Comment by redhog** [comment]
Heyas @mmaelicke ! Would you have time to review this?

Note: There is a new flag "sparse" added. If set to false, all distances are calculated using pdist, which is the fastest way, at least for semi-small datasets, but quickly eats a lot of ram (N*M, where N is number of points kriged to and M is number of points kriged from). If set to true, distances are calculated using ckDTree only for points withing range, and stored in a sparse matrix. This takes considerably less storage, but it slightly slower for lookup, so for smaller datasets this is a disadvantage. For larger datasets this saves your machine from swapping, which would quickly lower your performance.

**Comment by mmaelicke** [comment]
@redhog , awesome, thank a lot! 
Yeah, I will review it. I'll do my very best to do it this evening.

**Comment by redhog** [comment]
Maybe the default for sparse needs to be False... Hm...

File: skgstat/Kriging.py
Comment (mmaelicke) [inline skgstat/Kriging.py:33]: Can we rename it to `_sparse_dok_get`, as this is not intended to be exported from the module, right?
And maybe a one-liner docstring might be helpful

File: skgstat/Kriging.py
Comment (mmaelicke) [inline skgstat/Kriging.py:50]: I would plead for a `False` as the default value here. First, a `True` would change the current behavior. Second, I feel like the use cases in which `sparse=False` makes more sense occur more often. But I might be wrong in that feeling ;) 
We could also keep it `False`for now and turn it to `True` with the next minor version shift, along with a warning that will be raised until then.

File: skgstat/Kriging.py
Comment (mmaelicke) [inline skgstat/Kriging.py:428]: just out of curiosity: why stable sort here?

**Review by mmaelicke — CHANGES_REQUESTED** [review]
Nice work, thanks a lot!
I think, we need to sort the default value for `sparse`out, then I'm completely fine with it.

File: skgstat/Kriging.py
Comment (redhog) [reply skgstat/Kriging.py:428]: Because without that I couldn't do a regression test that gave the same result with sparse/non-sparse (when point coords coincided exactly for some pair of points, which some did in my test dataset)..

File: skgstat/Kriging.py
Comment (redhog) [reply skgstat/Kriging.py:50]: Hm, that's probably a good idea. Will push a change.

File: skgstat/Kriging.py
Comment (redhog) [reply skgstat/Kriging.py:33]: Sure. It's just a utility function.

**Comment by redhog** [comment]
This is superseded by https://github.com/mmaelicke/scikit-gstat/pull/68 which implements the same feature, but in a separate class that can also be used by Variogram.

**Comment by mmaelicke** [comment]
> This is superseded by #68 which implements the same feature, but in a separate class that can also be used by Variogram.

So, you want to only merge #68 instead of this PR, or does #68 just replace the utility function after this PR was merged?

**Comment by redhog** [comment]
Only merge the other one. Sorry for coding faster than I talked to you...

**Comment by redhog** [comment]
I had the other one in parallel, but didn't want to push it due to a bug that I have now squished...

**Comment by mmaelicke** [comment]
NP. I will just wait until you want me to review or merge something. Or both. :)

**Pull request closed without merging**
